src/MDMTransaction.m

```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target named:(NSString *)name {
+  if (name.length && [self planForName:name withTarget:target]) {
+    [self commonRemovePlan:plan fromTarget:target named:name];
+  }
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  [self commonRemovePlan:[self planForName:name withTarget:target] fromTarget:target named:name];
```

Comment:
`[self planForName:name withTarget:target]` is building more assumptions into the Transaction object than it needs to be responsible for.

A Transaction is often an ephemeral instance created at any time and containing any number of requested operations. A Transaction might look like this:

```
let transaction = Transaction()
transaction.removePlan(named:"someName")
scheduler.commit(transaction: transaction)
```

In this scenario the transaction understandably wouldn't know which plan is associated with a given name. The code as written has an expectation that the transaction is aware of plans registered to a given target, rather than simply relaying "remove a plan with a given name" to the scheduler.

To emphasize previous discussions about the transaction object: it is meant to be a very dumb vessel that captures requested operations and forwards those operations to the scheduler. The transaction has no long-term awareness of ongoing state; it solely enables communication from the application code to the scheduler object.

- [ ] The Transaction object's public APIs should simply be logged as "operations" in an ordered array, without any form of optimization or other state manipulation.
- [ ] When a transaction is committed to the scheduler, the scheduler enumerates each of these operations as though they were invoked on the scheduler itself.